Add single product block

thealexandrelara opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes #8568  

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Log in to your WordPress dashboard;
2. Go to Appearance > Themes, and select a block theme (for example: Twenty-twenty three);
3. Go to Appearance < Site Editor (Beta);
4. Click the Edit button;
5. Click on the "+" icon to add a new block and search for "Single Product" block in the search bar;
6. Click on the "Single Product" block to add it to your page or post;
7. Select a product on the Product Selector and click on the Save button;
8. Go to your website and check the selected product is being displayed with its correct information

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

### Changelog

> Add Single Product block that allows merchant to select and display a single product on their store

thealexandrelara committed:
Add block minimal structure

thealexandrelara committed:
Add Single Product block icon

thealexandrelara committed:
Add usesContext to block.json

thealexandrelara committed:
Add Save function to Single Product Block registration settings

thealexandrelara committed:
Add InnerBlocks template to Editor for Single Product Block

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Add Product selector to Single Product block

thealexandrelara committed:
Add withProduct HOC to the block editor

thealexandrelara committed:
Add save function to the Single Product block

thealexandrelara committed:
Add renderOnServerSide attribute to Product image

This attribute is used to determine if the product image should be rendered on the server side or not. This is needed because the product image is not rendered on the server side by default, but we want to render it on the server side when it is inside the Single Product block.

thealexandrelara committed:
Add renderOnServerSide attribute to Product Category List block

thealexandrelara committed:
Add renderOnServerSide attribute to Product Price block

thealexandrelara committed:
Add renderOnServerSide attribute to Product Rating block

thealexandrelara committed:
Add renderOnServerSide attribute to Product Stock Indicator block

thealexandrelara committed:
Add renderOnServerSide attribute to Product Summary block

thealexandrelara committed:
Add renderOnServerSide attribute to Product Tag List block

thealexandrelara committed:
Add renderOnServerSide attribute to Product Title block

thealexandrelara committed:
Add renderOnServerSide attribute to Product Add to Cart, Sales Badge and Sku block

thealexandrelara committed:
Reuse editor code for Single Product block

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Add Props interface to Single Product Server Side Editor

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Add render_block_context hook to SingleProductServerSide block

thealexandrelara committed:
Fix Single Product Server Side block that was not appearing in the block inserter

thealexandrelara committed:
Fix bug where the correct product was not being displayed on the frontend

gigitux committed:
Remove Single Product Block

gigitux committed:
Merge branch 'trunk' into fix/remove-single-product-block

thealexandrelara committed:
Add Product title and product summary variations

thealexandrelara committed:
Add Add To Cart Form and Product Meta blocks to the Single Product Block

gigitux committed:
remove more files that belong the Single Product block

gigitux committed:
Merge branch 'fix/remove-single-product-block' of https://github.com/woocommerce/woocommerce-blocks into fix/remove-single-product-block

gigitux committed:
fix php linter error

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/remove-single-product-block

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

gigitux committed:
Merge branch 'fix/remove-single-product-block' into feat/add-single-product-block-server-side

gigitux committed:
wrap up Single Product Block

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Fix issue preventing the block from working on the Single Product Template

thealexandrelara committed:
Fix update context function for Single Product block

thealexandrelara committed:
Replace Single Product block title

thealexandrelara committed:
Replace global $post only for Single Product inner blocks

thealexandrelara committed:
Remove unnecessary var_dump statement

thealexandrelara committed:
Improve documentation for the Single Product block

thealexandrelara committed:
Improve documentation for Single Product block

gigitux committed:
fix registration: add-to-cart-form and product-meta blocks

thealexandrelara committed:
Fix lint error related to imports

thealexandrelara committed:
Remove query monitor debug code

thealexandrelara committed:
Remove unnecessary `renderOnServerSide` attribute from product elements

thealexandrelara committed:
Fix return type for the Save function of the Product Elements blocks

thealexandrelara committed:
Fix return type for the Save function of the Product Title block

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Add scss file for Single Product block

thealexandrelara committed:
Fix PHP coding standards

thealexandrelara committed:
Fix php coding standards

Aljullu commented on assets/js/blocks/single-product/index.tsx:
```diff
@@ -0,0 +1,22 @@
+ */
+import { registerBlockType } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { BLOCK_ICON } from './constants';
+
+/**
+ * Internal dependencies
+ */
+import metadata from './block.json';
+import edit from './edit';
+import save from './save';
```
Internal dependencies is repeated.

thealexandrelara replied:
Nice catch! I removed the duplication

Aljullu commented on assets/js/blocks/single-product/constants.tsx:
```diff
@@ -0,0 +1,79 @@
+				{},
+				[
+					[
+						'core/post-title',
+						{
+							headingLevel: 2,
+							__woocommerceNamespace:
+								PRODUCT_TITLE_VARIATION_NAME,
+						},
```
I don't have a strong opinion, but was it considered making the Product Title a link by default? Currently it isn't.

![imatge](https://user-images.githubusercontent.com/3616980/226612758-c26c4f27-a8f2-4a15-a0c0-4278d0caa2a8.png)

thealexandrelara replied:
I agree with you, I think that having the title to, by default, be a link would be a much better user experience. At least, it seems natural to click on the product title to go to the product page

Aljullu commented on assets/js/blocks/single-product/edit/editor-block-controls.tsx:
```diff
@@ -0,0 +1,43 @@
+}: EditorBlockControlsProps ) => {
+	return (
+		<BlockControls>
+			<ToolbarGroup
```
For some reason, the toolbar is not visible to me. Does it work for you? :confused: 

![imatge](https://user-images.githubusercontent.com/3616980/226613710-4aa74728-d997-4c77-8253-f4a9966c26f4.png)

thealexandrelara replied:
Nice catch! You're right, the Block Control was not being displayed. I noticed that the `blockProps` was not being added to the wrapper component of the block

Aljullu commented on assets/js/blocks/single-product/edit/editor.scss:
```diff
@@ -0,0 +1,18 @@
+	.wc-block-single-product__edit-card-title {
+		margin: 0 0 $gap;
+	}
+}
```
For class names which are only visible in the editor, we [used to prefix them with `.wc-block-editor-`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/contributors/contributing/coding-guidelines.md#css-class-names). Do you think we should change these ones?

thealexandrelara replied:
I agree, just changed them!

Aljullu commented on assets/js/blocks/single-product/edit/layout-editor.tsx:
```diff
@@ -0,0 +1,97 @@
+							isTertiary
+							className="wc-block-single-product__reset-layout"
+						>
+							<Icon icon={ backup } />{ ' ' }
```
Here you could use the [`icon` prop](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/button#icon-iconprops-unknown--icon-) which is available for the `Button` component. This way you don't need to import the `Icon` component. Then you can also remove the `svg` rules from `assets/js/blocks/single-product/edit/editor.scss`.

thealexandrelara replied:
Oh this is nice, much better to use the prop from the Button component. Thank you!

Aljullu reviewed (commented):
This looks really cool, good job, @thealexandrelara! I did some testing and overall everything seemed to work quite smoothly. Leaving some comments below with what I could find:

1. Adding to Cart redirects to the product page. Is that expected?
2. In TT3, the Add to Cart input field appears very small. Can you reproduce as well?

![imatge](https://user-images.githubusercontent.com/3616980/226613069-c6e30959-b2b6-4590-becf-f4e9fe243b55.png)

3. I saw there are some PHP coding standards errors that might need to be fixed.
4. I left some other comments inline below. :point_down:

thealexandrelara committed:
Remove internal dependencies section duplication

thealexandrelara committed:
Replace .wc-block with .wc-block-editor- styles

According to the naming convention for this project (https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/contributors/contributing/coding-guidelines.md#naming) we should be using the `.wc-block-editor-` prefix for styles that are only used in the editor.

thealexandrelara committed:
Solve style for the Edit card component in the Single Product block

thealexandrelara committed:
Fix error preventing Block Control from being displayed

thealexandrelara committed:
Make the Product title to be a link by default

thealexandrelara committed:
Fix PHP Coding Standards

thealexandrelara committed:
Fix PHP Code Standards

thealexandrelara committed:
Fix php cs issues

thealexandrelara committed:
Fix issue with php cs

thealexandrelara committed:
Remove unnecessary comment

thealexandrelara commented:
Thank you for reviewing this PR @Aljullu!

> 1. Adding to Cart redirects to the product page. Is that expected?

For now, I would say yes, but in the future this is something we need to address, I opened [this issue](https://github.com/woocommerce/woocommerce-blocks/issues/8771) a few days ago so we can start discussing some possible solutions.  Maybe I'm wrong but I think that currently the behavior is being controlled by WooCommerce and not by the Add To Cart Form block itself :/

> 2. In TT3, the Add to Cart input field appears very small. Can you reproduce as well?

I think this is something we can address in the same issue that I mentioned above because it's kind of related. Initially, the Add to Cart Form block was created considering that it would work for the Single Product Template, but now that we can use in multiple places, some styles are not being applied when outside the Single Product Template.

> 3. I saw there are some PHP coding standards errors that might need to be fixed.

Just fixed them, there is just one related to the global $post that it's basically the one way we found to make the correct product to display for this block, so I cannot edit it

thealexandrelara committed:
Remove unused Icon import

nefeline commented on src/BlockTypes/AddToCartForm.php:
```diff
@@ -25,19 +25,27 @@ class AddToCartForm extends AbstractBlock {
+
+		$product = wc_get_product( $post_id );
 
-		while ( have_posts() ) {
```
Hi there! Do you mind sharing more details regarding this change? For the same reasons being discussed on https://github.com/woocommerce/woocommerce-blocks/discussions/8792#discussioncomment-5387577 , I would not recommend stop relying on `while ( have_posts() ) { the_post(); }`

thealexandrelara replied:
Hey @nefeline, I answered in the discussion post you shared, but just bringing an overview here. That was necessary because now the Single Product block can work in different contexts, for instance, in the Home template. Also, the merchants can select any product to be displayed and not only the product of the page they are accessing.

nefeline replied:
Gotcha, thanks for clarifying! I think we can explore a different approach here: what do you think about [this proposal](https://github.com/woocommerce/woocommerce-blocks/pull/8610/files#r1148088334) ?

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

Aljullu commented on assets/js/blocks/single-product/edit/index.tsx:
```diff
@@ -0,0 +1,147 @@
+					<Placeholder
+						icon={ BLOCK_ICON }
+						label={ metadata.name }
+						className=".wc-block-editor-single-product"
```
It looks like in some places the dot got inside the `className` prop. It might be good to do a search for `className=".` across the codebase.

thealexandrelara replied:
Just removed it! 🙌

Aljullu commented on assets/js/blocks/single-product/edit/index.tsx:
```diff
@@ -0,0 +1,147 @@
+				{ isEditing ? (
+					<Placeholder
+						icon={ BLOCK_ICON }
+						label={ metadata.name }
```
This should be `metadata.title`:
```suggestion
						label={ metadata.title }
```

otherwise, the slug is displayed:

![imatge](https://user-images.githubusercontent.com/3616980/227502288-39fa4e37-846b-40ba-857f-4cf4130a2c7c.png)

Aljullu commented on src/BlockTypes/SingleProduct.php:
```diff
@@ -0,0 +1,139 @@
+			$block_name = array_pop( $this->single_product_inner_blocks_names );
+
+			if ( $block_name === $block['blockName'] ) {
+				global $post;
+				$post = get_post( $this->product_id );
```
> Just fixed them, there is just one related to the global $post that it's basically the one way we found to make the correct product to display for this block, so I cannot edit it

In that case, I think we should add a `phpcs:ignore` directive so the automated job keeps passing.

Aljullu reviewed (commented):
Thanks for updating this PR, @thealexandrelara! This is looking great! There are still some edge cases I found, listing them below: :point_down: 

1. Not sure if it needs to be part of this PR, but mentioning in case it's not known: the Product Meta block doesn't seem to work on the Single Product block:

Editor | Frontend
--- | ---
![imatge](https://user-images.githubusercontent.com/3616980/227502997-3682998a-0821-4712-b56e-95a65f094168.png) | ![imatge](https://user-images.githubusercontent.com/3616980/227503055-3d5dd05b-12c1-4c74-beef-c7aa165bc407.png)


2. We should review the `TypeScript Errors Report`, as it seems like this PR is introducing some new errors? :thinking: 
3. Not a blocker, but has been considered to support wide and full-width alignments?

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Remove unnecessary dot from the className attributes in the Single Product Block

thealexandrelara committed:
Replace metadata.name with metadata.title in the Editor file

thealexandrelara committed:
Ignore phpcs rule

thealexandrelara committed:
Add new attributes to the Attributes interface of the Price block

thealexandrelara committed:
Fix typescript errors on the Product Meta block

thealexandrelara committed:
Fix ts errors on Single Product editor

thealexandrelara committed:
Fix ts errors on layout-editor.tsx

thealexandrelara committed:
Fix ts errors in Single Product index file

thealexandrelara committed:
Fix ts errors on add to cart form index file

thealexandrelara commented:
Thank you for the detailed review!

> 1. Not sure if it needs to be part of this PR, but mentioning in case it's not known: the Product Meta block doesn't seem to work on the Single Product block:

@gigitux I know that you're working on #8805, do you know if it would fix on the frontend side as well? Or it is better to create other PR to handle this?

> 2. We should review the `TypeScript Errors Report`, as it seems like this PR is introducing some new errors? 🤔

I just reviewed them and solved most of them, some are being caused by external dependencies, in this case should we just ignore them?

> 3. Not a blocker, but has been considered to support wide and full-width alignments?

That would be a great addition, I'll open an issue and work on this after merging this PR.

nefeline reviewed (commented):
Nice work here @thealexandrelara 🎉 !

During the tests, I noticed a few things:

- When adding the "Single Product" block to the Header template part, the add to cart and product details blocks are not always rendered:  

<img width="1231" alt="Screenshot 2023-03-24 at 18 09 06" src="https://user-images.githubusercontent.com/15730971/227642213-23d66bc7-b324-412d-9215-1a0deb23b160.png">

In some attempts, the addition was successful, while in others it wasn't.

- No preview is available for this block: is that expected at this point?

<img width="504" alt="Screenshot 2023-03-24 at 18 21 34" src="https://user-images.githubusercontent.com/15730971/227644008-533b230a-733b-4abd-af04-a642feca126b.png">

- There's a 404 error for the `single-product-frontend.js`: I also noticed the same inconsistency that @Aljullu did ref. the quantity selector:

<img width="1234" alt="Screenshot 2023-03-24 at 18 26 22" src="https://user-images.githubusercontent.com/15730971/227644894-f59f2ff2-ce3c-484f-9438-927c9327c3c2.png">

- Although it is properly displayed when in the context of a single product template or the shop:

<img width="1234" alt="Screenshot 2023-03-24 at 18 31 33" src="https://user-images.githubusercontent.com/15730971/227645397-079599ae-64f8-41e4-bd4f-bc4c11fe06a8.png">

nefeline commented on src/BlockTypes/SingleProduct.php:
```diff
@@ -0,0 +1,140 @@
+			if ( $block_name === $block['blockName'] ) {
+				global $post;
+				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+				$post = get_post( $this->product_id );
```
I would encourage finding an alternative approach here rather than overriding the WordPress `$post` global variable for a few reasons:

* Overwriting WordPress global variables can cause conflicts and unexpected behavior.
* When we overwrite global variables, there's a risk of breaking core functionality or interfering with the functionality of plugins and themes that rely on those. 
* Additionally, it can make the code harder to maintain and debug, as it can be challenging to track down where a particular value is being set or changed.

As an alternative, how about updating the `render` method for each one of the relevant blocks, so they have different behaviors depending on the context? Assuming the `$product_id` value for the block can be fetched from the DB:

example for the Add to cart form:

```php
ob_start();
// When in the context of the single product template, behave exactly as on Woo core:
if ( is_product() ) {
	while ( have_posts() ) {
		the_post();
		global $product;
		do_action( 'woocommerce_' . $product->get_type() . '_add_to_cart' );
	}
} else {
        // When used anywhere else on the site, fetch the saved product ID selected by the user for this block instead:
	$product = wc_get_product( $product_id );

	if ( ! $product ) {
		return;
	}

	do_action( 'woocommerce_' . $product->get_type() . '_add_to_cart' );
}

$product = ob_get_clean();
```

What do you think?

thealexandrelara replied:
I agree that this is not the best approach and thank you for suggesting a new one; however, I think this could be handled in a separate PR that would refactor every block to use the approach that we decide on [Using the Blockified Single Product Template](https://github.com/woocommerce/woocommerce-blocks/discussions/8792). Editing the `render` method would not work for the variations blocks, because we cannot edit their render method :/ (ref: `per0F9-qM-p2`)

nefeline replied:
Thank you so much for sharing the link to that discussion and more context about this problem! I added a comment over there. It looks like the only block that is relying on the global post variable is the Post Excerpt. Cross-posting my thoughts over here:

I was taking a look at the proposed solution, which currently is [this one](https://href.li/?https://github.com/woocommerce/woocommerce-blocks/blob/675f3b176edcc84e7bce771a2afd5b8b4fcd45d8/src/BlockTypes/SingleProduct.php#L123-L140):

```php
         protected function replace_post_for_single_product_inner_block( $block, &$context ) {
		if ( $this->single_product_inner_blocks_names ) {
			$block_name = array_pop( $this->single_product_inner_blocks_names );

			if ( $block_name === $block['blockName'] ) {
				global $post;
				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
				$post = get_post( $this->product_id );
				setup_postdata( $post );
				$context['postId'] = $this->product_id;
			}

			if ( ! $this->single_product_inner_blocks_names ) {
				wp_reset_postdata();
			}
		}
	}
```

I did a small experiment as while I did understand why we needed to update the `$context['postId']` , I didn’t understand why we needed to overwrite the post global variable. The above method was updated to:

```php
	protected function replace_post_for_single_product_inner_block( $block, &$context ) {
		if ( $this->single_product_inner_blocks_names ) {
			$block_name = array_pop( $this->single_product_inner_blocks_names );

			if ( $block_name === $block['blockName'] && ! is_product() ) {
				$context['postId'] = $this->product_id;
			}
		}
	}
```

With this change in place, the Single Product block worked as expected, and all inner blocks were rendered, except for the Product Summary block, which is a variation of the Post Excerpt block. Next, I checked what’s going on on the Gutenberg side, [this is what renders the post excerpt](https://href.li/?https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-excerpt/index.php#L29-L31) content:

```php
function render_block_core_post_excerpt( $attributes, $content, $block ) {
	if ( ! isset( $block->context['postId'] ) ) {
		return '';
	}

	/*
	* The purpose of the excerpt length setting is to limit the length of both
	* automatically generated and user-created excerpts.
	* Because the excerpt_length filter only applies to auto generated excerpts,
	* wp_trim_words is used instead.
	*/
	$excerpt_length = $attributes['excerptLength'];
	if ( isset( $excerpt_length ) ) {
		$excerpt = wp_trim_words( get_the_excerpt(), $excerpt_length );
	} else {
		$excerpt = get_the_excerpt();
	}
	$more_text           = ! empty( $attributes['moreText'] ) ? '<a class="wp-block-post-excerpt__more-link" href="' . esc_url( get_the_permalink( $block->context['postId'] ) ) . '">' . wp_kses_post( $attributes['moreText'] ) . '</a>' : '';
	$filter_excerpt_more = function( $more ) use ( $more_text ) {
		return empty( $more_text ) ? $more : '';
	};
	/**
	 * Some themes might use `excerpt_more` filter to handle the
	 * `more` link displayed after a trimmed excerpt. Since the
	 * block has a `more text` attribute we have to check and
	 * override if needed the return value from this filter.
	 * So if the block's attribute is not empty override the
	 * `excerpt_more` filter and return nothing. This will
	 * result in showing only one `read more` link at a time.
	 */
	add_filter( 'excerpt_more', $filter_excerpt_more );
	$classes = array();
	if ( isset( $attributes['textAlign'] ) ) {
		$classes[] = 'has-text-align-' . $attributes['textAlign'];
	}
	if ( isset( $attributes['style']['elements']['link']['color']['text'] ) ) {
		$classes[] = 'has-link-color';
	}
	$wrapper_attributes = get_block_wrapper_attributes( array( 'class' => implode( ' ', $classes ) ) );

	$content               = '<p class="wp-block-post-excerpt__excerpt">' . $excerpt;
	$show_more_on_new_line = ! isset( $attributes['showMoreOnNewLine'] ) || $attributes['showMoreOnNewLine'];
	if ( $show_more_on_new_line && ! empty( $more_text ) ) {
		$content .= '</p><p class="wp-block-post-excerpt__more-text">' . $more_text . '</p>';
	} else {
		$content .= " $more_text</p>";
	}
	remove_filter( 'excerpt_more', $filter_excerpt_more );
	return sprintf( '<div %1$s>%2$s</div>', $wrapper_attributes, $content );
}
```

Note how `get_the_excerpt()` is invoked without receiving the post_id as a variable. As a consequence, it attempts to fetch the excerpt from the global context (which will be the post where the block was added instead of being based on the block context).

In other words, with a change as simple as `get_the_excerpt( $block->context['postId'] );` to the `render_block_core_post_excerpt` function, we can achieve all goals without overwriting any global variables.

With that said, what are your thoughts about:

1. Proposing this change to the core of Gutenberg
2. Relying on a dedicated version for the Post Excerpt in the meantime

thealexandrelara replied:
Thank you for bringing this solution here, yes I agree with you we can implement your suggestion and it would be much better. I'll be changing the code to match what you suggested and put a conditional to only apply the global post when it is the post excerpt block, at least until GB approves the change to its core, or we have a dedicated version of the Post Excerpt

thealexandrelara replied:
@nefeline I replaced `replace_post_for_single_product_inner_block` with your suggested solution, but it gives me an error when I go to the frontend, am I missing something?

<img width="1227" alt="image" src="https://user-images.githubusercontent.com/20469356/228928783-2c6d29ef-dbb0-40bc-88a5-e50e3f0655f9.png">

nefeline replied:
Sorry about that! I forgot to share a few changes I did to the AddToCartForm to support this; they are [available here](https://github.com/woocommerce/woocommerce-blocks/compare/feat/add-single-product-block-server-side...try/global-variable-overwrite?expand=1).

thealexandrelara replied:
Thank you for sharing this! Now the error is gone, but I noticed that if I add the block inside the Single Product template and visit a product page, the product that is displayed is not the one that was selected in the Editor, is the product that the user is currently visiting

nefeline replied:
> I noticed that if I add the block inside the Single Product template and visit a product page, the product that is displayed is not the one that was selected in the Editor, is the product that the user is currently visiting

Nice catch! That's because of [this line](https://github.com/woocommerce/woocommerce-blocks/blob/576cc4151bf1587453f9f86d187768debb3ab2b5/src/BlockTypes/SingleProduct.php#L127): if we ditch the `! is_product()` validation we are all set: in reality, it doesn't matter if we are in the context of a single product template or not; we will always want to fetch the content from the block context.

thealexandrelara commented:
Thank you for taking a look on this and providing such a detailed review! 🙌 

> Nice work here @thealexandrelara 🎉 !
> 
> During the tests, I noticed a few things:
> 
> * When adding the "Single Product" block to the Header template part, the add to cart and product details blocks are not always rendered:
> 
> In some attempts, the addition was successful, while in others it wasn't.

Nice catch! @gigitux worked on a solution to make these blocks available in different contexts (templates, posts, pages, etc) but I don't know if it was meant to work inside the template parts. Since he is currently unavailable I'll be taking a look on this

> * No preview is available for this block: is that expected at this point?

I would say yes, but I can open a new issue to work on this in a separate PR

> * There's a 404 error for the `single-product-frontend.js`: I also noticed the same inconsistency that @Aljullu did ref. the quantity selector:
> * Although it is properly displayed when in the context of a single product template or the shop:

I opened up a separate [issue](https://github.com/woocommerce/woocommerce-blocks/issues/8771) to handle this because it's related to the Add To Cart Form block. About the 404 error, I can't reproduce it, but since we deleted the previous single product block, maybe you're facing some cache issue, have you tried on Incognito mode?

thealexandrelara committed:
Fix unsupported block error when using inner blocks outside the Single Product template

thealexandrelara committed:
Fix unsupported block error for the Single Product block when editing Posts/Pages

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

Aljullu commented on assets/js/atomic/blocks/product-elements/product-image-gallery/edit.tsx:
```diff
@@ -58,7 +58,7 @@ const Edit = ( { context }: Props ) => {
 		);
 	}
 	// We have work on this case when we will work on the Single Product block.
```
Should this comment be removed, now? Or what's missing here?

thealexandrelara replied:
Luigi was working on this component, I think he mentioned that this comment was there because in the future we could think about displaying the image for the selected product instead of the image placeholder; perhaps we should create an issue and remove the comment? 🤔

Aljullu replied:
cc'ing @gigitux for his thoughts on this. :point_up:

gigitux replied:
At least, for now, we decided to use the `Product Image` block for the Single Product Block. In the future, we should add support for the `Product Image Gallery`. (we don't have a dedicated roadmap) - related discussion ( p1678382112656919-slack-C02FL3X7KR6).
It makes sense to open an issue and link it in this comment. What do you think?

Aljullu commented on assets/js/blocks/single-product/index.tsx:
```diff
@@ -0,0 +1,19 @@
+import edit from './edit';
+import save from './save';
+
+// @ts-expect-error: `registerBlockType` is a function that is typed in WordPress core.
```
I know this comment is based on code prior to this PR, but to me it's not clear why there is a TS error here. `registerBlockType` is a function that is typed in WordPress core, but why do we have to add `@ts-expect-error`? Sorry if it's something obvious that I'm missing. :sweat_smile:

thealexandrelara replied:
Well, I think this became a "default" at some point and now it is being copied-pasted in multiple places 😅 . The `registerBlockType` accepts a string or an [object as its first parameter](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#javascript-client-side); however, in our project the type for the first parameter is just a string which causes the error when we use the `metadata` object that comes from the "block.json" file

Aljullu replied:
Yeah... no need to fix it in this PR, but it would be good to follow-up and improve it. :slightly_smiling_face:

gigitux replied:
The function `registerBlockType` isn't typed well. So, we should improve it upstream (not sure it is easy). The comment is useful because if it is fixed, a TS error will be shown and we can remove the comment. (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments)

Aljullu commented on assets/js/blocks/single-product/edit/index.tsx:
```diff
@@ -0,0 +1,148 @@
+	const [ isEditing, setIsEditing ] = useState( ! productId );
+	const blockProps = useBlockProps();
+
+	if ( isPreview ) {
+		return singleProductBlockPreview;
+	}
```
Should something render as the block preview in the inserter? It doesn't seem to work for me. :confused: 

![imatge](https://user-images.githubusercontent.com/3616980/228863693-43e27349-aac8-4ea1-af45-184384091136.png)

thealexandrelara replied:
I'm going to create a separate PR to solve that because this part is from the old single product block, as soon as we merge this PR I'll start working on the edge cases and enhancements. I put this behind the experimental flag as you suggested

Aljullu replied:
Sounds good. :+1:

Aljullu commented on assets/js/blocks/single-product/edit/api-error.tsx:
```diff
@@ -0,0 +1,31 @@
+ * @param {boolean}           props.isLoading
+ * @param {function(any):any} props.getProduct
+ */
+const ApiError = ( { error, isLoading, getProduct }: ApiErrorProps ) => (
```
Not a big deal, but given that this component is only used once and its only logic is to add a classname, I wonder if it's really needed or we could just call `ErrorPlaceholder` directly.

thealexandrelara replied:
Yes, it is not needed, just removed it! It was being used in the old Single Product block implementation

Aljullu commented on assets/js/blocks/single-product/edit/index.tsx:
```diff
@@ -0,0 +1,148 @@
+/**
+ * Component to handle edit mode of the "Single Product Block".
+ *
+ * @param {Object}            props               Incoming props for the component.
+ * @param {string}            props.className
+ * @param {Object}            props.attributes    Incoming block attributes.
+ * @param {function(any):any} props.setAttributes Setter for block attributes.
+ * @param {string}            props.error
+ * @param {function(any):any} props.getProduct
+ * @param {Object}            props.product
+ * @param {boolean}           props.isLoading
+ * @param {string}            props.clientId
```
Not a blocker either, but given that this component is typed with TS, could we skip the JSDoc types? Or is it needed for something else?

thealexandrelara replied:
You're correct. These files inside the `edit` folder was copied from the old Single Product block and they were JS files so we missed these annotations in here, thank you for reporting that

Aljullu commented on assets/js/blocks/single-product/edit/editor-block-controls.tsx:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Adds controls to the editor toolbar.
+ *
+ * @param {Object}                props              Incoming props for the component.
+ * @param {boolean}               props.isEditing
+ * @param {function(boolean):any} props.setIsEditing
```
Same here. :slightly_smiling_face:

Aljullu reviewed (commented):
This is shaping up nicely. Good job, @thealexandrelara! I could still find some rough edges, though. I wonder if we should feature-gate this block to experimental or the feature plugin for now. That would ease getting this PR merged and work on some of the things we find in follow-ups. What do you think?

Some things I found:

* Product Summary variation is crashing in the editor for me (WP 6.2, GB 15.4). Can you reproduce as well? I also see there is a JS error, which might be related: `TypeError: can't access property "trim", x is undefined`.

![Captura de pantalla de 2023-03-30 15-59-42](https://user-images.githubusercontent.com/3616980/228860812-dce0b133-3719-4907-9365-f74664feb20f.png)

* I also noticed that the product image has different sizing in the editor and the frontend. I guess that's because in the frontend we are rendering a thumbnail and in the editor the original (or something like that)? Ideally we should unify them so they render the same:

Editor | Frontend
--- | ---
![imatge](https://user-images.githubusercontent.com/3616980/228861815-0ea0dccd-de21-4993-9066-2376957974c3.png) | ![imatge](https://user-images.githubusercontent.com/3616980/228862037-9aefef58-2b51-4739-be57-c0df629075d5.png)

nefeline reviewed (commented):
> Nice catch! @gigitux worked on a solution to make these blocks available in different contexts (templates, posts, pages, etc) but I don't know if it was meant to work inside the template parts. Since he is currently unavailable I'll be taking a look on this

👍 sounds good

> I would say yes, but I can open a new issue to work on this in a separate PR

👍 

> I opened up a separate https://github.com/woocommerce/woocommerce-blocks/issues/8771 to handle this because it's related to the Add To Cart Form block. 

Thanks! From a styling perspective (and not the redirect), I would say it's important to fix this in the context of this PR, as it is a bit of a bummer to add the block to a post/page outside of the context of a single product and be presented to something like this:

<img width="922" alt="Screenshot 2023-03-30 at 11 41 19" src="https://user-images.githubusercontent.com/15730971/228873496-039af212-ffd5-49f9-9a25-9ef9f2a965e3.png">

> About the 404 error, I can't reproduce it, but since we deleted the previous single product block, maybe you're facing some cache issue, have you tried on Incognito mode?

Odd! There's not a single scenario where I can't reproduce this error; every time the block is added to any post/page, the 404 error is there:

<img width="1233" alt="Screenshot 2023-03-30 at 11 48 59" src="https://user-images.githubusercontent.com/15730971/228875688-b0911287-bc74-4935-841b-fffaccc820a1.png">

I'm glad to jump on a call and/or screen share if needed!

Edit: I just realized I forgot to answer your questions regarding the 404, my apologies! I have deleted the block and started all over from scratch as suggested: still can reproduce the error, even in Incognito mode.

thealexandrelara committed:
Remove unnecessary JSDoc types

thealexandrelara committed:
Remove unnecessary APIError component

thealexandrelara commented:
>  I wonder if we should feature-gate this block to experimental or the feature plugin for now. 

I totally agree with you, I think that it's better to merge this PR behind a experimental flag and handle all of the edge cases and issues being caused by its inner blocks

> * Product Summary variation is crashing in the editor for me (WP 6.2, GB 15.4). Can you reproduce as well? I also see there is a JS error, which might be related: `TypeError: can't access property "trim", x is undefined`.

Yeah, I can confirm that this happens sometimes. I just saw that they merged a fix on Gutenberg last week: https://github.com/WordPress/gutenberg/pull/48730

> I also noticed that the product image has different sizing in the editor and the frontend. I guess that's because in the frontend we are rendering a thumbnail and in the editor the original (or something like that)? Ideally we should unify them so they render the same:

Nice catch! I'll be investigate more on this and open an issue if necessary

thealexandrelara committed:
Replace global $post only when rendering a `core/post-excerpt` block

thealexandrelara committed:
Fix php cs errors

thealexandrelara committed:
Put the Single Product block behind an experimental flag

thealexandrelara committed:
Fix error occurring when trying to access ancestor of an undefined block settings

thealexandrelara commented:
Hey @Aljullu and @nefeline, thank you for reviewing this PR!
Since there are many issues related to the inner blocks being used by this block, I took what Albert suggested and put this block behind the Experimental flag, this way we can leave this PR to be scoped to display the information for the selected product using the inner blocks, and then we can work on solving the issues introduced by the inner blocks on separate PRs, which would be easier to review and preventing us from having a huge PR that solves everything.

@Aljullu  I opened the issue https://github.com/woocommerce/woocommerce-blocks/issues/8925 to work on what you reported about the image size being different depending on where the user is (on the Editor or on the Frontend), thank you for raising this.

@nefeline I implemented your suggestion for avoiding the usage of the global variables, but I had to temporarily replace the global $post for the Post Excerpt block until we decide the best way to handle this (by waiting Gutenberg merge your PR or by creating a "ghost" block)

Considering that, I think the best approach now would be to map the issues related to the inner blocks, keep this block behind the experimental flag and merge it (if we don't have any issue being caused by the Single Product block), and start working on the raised issues in separate PRs. When we finish solving them we can create a PR to remove the experimental flag of this block and test it deeper, what do you think?

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

nefeline commented:
Thanks for all your work and also for addressing the feedback @thealexandrelara! Regarding:

> Considering that, I think the best approach now would be to map the issues related to the inner blocks, keep this block behind the experimental flag and merge it (if we don't have any issue being caused by the Single Product block), and start working on the raised issues in separate PRs. When we finish solving them we can create a PR to remove the experimental flag of this block and test it deeper, what do you think?

Unless we have a specific reason to release this sooner than later, perhaps another option we can explore is to park the work on this PR until we have all problems for each one of the inner blocks solved instead, so we can release it without any console errors or functionality problems. 

When we worked on blocks such as the Add to Cart Form, the Product Details, and others, the focus was to ensure they were all functional in the context of the single product template exclusively. Now we want them to be functional across the board, which requires extra work that I agree with you that could be done as part of separate PRs to avoid confusion and make more atomic changes. What are your thoughts?

Aljullu commented on assets/js/blocks/single-product/edit/shared-product-control.tsx:
```diff
@@ -0,0 +1,39 @@
+
+/**
+ * Allows a product to be selected for display.
+ *
+ * @param {Object}            props               Incoming props for the component.
+ * @param {Object}            props.attributes    Incoming block attributes.
+ * @param {function(any):any} props.setAttributes Setter for block attributes.
```
I would remove these JSDoc declarations here too. :slightly_smiling_face:

Aljullu commented on assets/js/atomic/utils/register-block-single-product-template.ts:
```diff
@@ -1,46 +1,96 @@
 		if (
-			block === undefined &&
-			parsedTemplate.includes( 'single-product' )
+			isBlockRegistered &&
+			( currentTemplateId?.includes( 'single-product' ) ||
+				previousTemplateId?.includes( 'single-product' ) )
 		) {
-			registerBlockFn();
+			if ( isVariationBlock && variationName ) {
+				unregisterBlockVariation( blockName, variationName );
+			} else {
+				unregisterBlockType( blockName );
+			}
+			isBlockRegistered = false;
 		}
```
I'm a bit confused by this code. Why is it needed to unregister the blocks here?

thealexandrelara replied:
The Single Product template is the only template where the inner blocks can exist outside the Single Product block because the template would provide the necessary context to the blocks. In this case, when the user visits the Single Product template, the block is registered without the `ancestor` key, but if the user leaves the Single Product page and goes to another template, for example, the Home template, the application will keep the block registered without the ancestor key, and the inner blocks like the Product Image, Product Price, etc. will be available in the Block Inserter which is not what we want

Unfortunately, I didn't find a way to dynamically modify the `ancestor` parameter of the block. Hence, the way we circumvent that was to unregister the block when the user enters or leaves the Single Product page, is it possible to change the ancestor key without unregistering and re-registering the blocks?

Aljullu replied:
Makes sense, thanks for explaining, @thealexandrelara. Do you think we could add a comment in that file with part of your explanation so anybody looking at the file knows its purpose?

thealexandrelara replied:
Nice, thank you for raising this! Just added the comment explaining this part of the code!

Aljullu reviewed (commented):
Thanks for updating this PR, @thealexandrelara!

For some reason, the Product Title isn't showing up for me when it's before the Product Summary block. Can you reproduce as well?

Editor | Frontend
--- | ---
![imatge](https://user-images.githubusercontent.com/3616980/229766481-1b5b4cc2-a590-4373-a535-67248eb49eae.png) | ![imatge](https://user-images.githubusercontent.com/3616980/229766532-ba1cb9bd-bfef-427f-a8fb-dd278cb5854d.png)

Btw, it might also be a good idea to rebase this PR on `trunk` to get e2e tests passing.

> Since there are many issues related to the inner blocks being used by this block, I took what Albert suggested and put this block behind the Experimental flag, this way we can leave this PR to be scoped to display the information for the selected product using the inner blocks, and then we can work on solving the issues introduced by the inner blocks on separate PRs, which would be easier to review and preventing us from having a huge PR that solves everything.

Sounds good to me.

> Unless we have a specific reason to release this sooner than later, perhaps another option we can explore is to park the work on this PR until we have all problems for each one of the inner blocks solved instead, so we can release it without any console errors or functionality problems.

@nefeline you mean leaving this PR open instead of merging it even if it's behing a feature flag? I would say the main advantage of merging early is to avoid this PR getting conflicts. Given that the block would be behind a feature gate, it should be ok, no? :thinking:

thealexandrelara committed:
Remove JS Doc from shared-product-control.tsx

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Fix single-product-frontend.js 404 error

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara commented:
> Unless we have a specific reason to release this sooner than later, perhaps another option we can explore is to park the work on this PR until we have all problems for each one of the inner blocks solved instead, so we can release it without any console errors or functionality problems.
> 
> When we worked on blocks such as the Add to Cart Form, the Product Details, and others, the focus was to ensure they were all functional in the context of the single product template exclusively. Now we want them to be functional across the board, which requires extra work that I agree with you that could be done as part of separate PRs to avoid confusion and make more atomic changes. What are your thoughts?

Thank you for pointing about the 404 error, after syncing with trunk I finally was able to reproduce it and I just solved by implementing the `get_block_type_script` method on the SingleProduct.php file 🙌 

I put the changes behind the Experimental flag so it won't be released until we decide to. The problem of parking this PR is that we would have to checkout branches having this one as base, keep them synced with this branch and also with trunk, so in the end we can merge with this one, which would make the PR get really big and hard to review. I think it would be better to rely on the experimental flag and keep working on each inner block separately until we feel it is finally ready to remove the flag and test everything together

nefeline commented:
> @nefeline you mean leaving this PR open instead of merging it even if it's behing a feature flag? I would say the main advantage of merging early is to avoid this PR getting conflicts. Given that the block would be behind a feature gate, it should be ok, no? 🤔

Yeah, that's what I meant, but you are right: merging early and keeping it behind a feature flag makes sense! 👍 

> Thank you for pointing about the 404 error, after syncing with trunk I finally was able to reproduce it and I just solved by implementing the get_block_type_script method on the SingleProduct.php file 🙌

Excellent, thanks for addressing the feedback! I can confirm this error is now solved.

> I put the changes behind the Experimental flag so it won't be released until we decide to. The problem of parking this PR is that we would have to checkout branches having this one as base, keep them synced with this branch and also with trunk, so in the end we can merge with this one, which would make the PR get really big and hard to review. I think it would be better to rely on the experimental flag and keep working on each inner block separately until we feel it is finally ready to remove the flag and test everything together

Gotcha, makes sense!

nefeline reviewed (changes requested):
Thanks for the updates here @thealexandrelara! This block is shaping up nicely.

I did a new round of tests:

1. When adding the Single Product block to a template, on the editor side, I get the following error: `Block names must be strings.`

<img width="1235" alt="Screenshot 2023-04-04 at 15 14 25" src="https://user-images.githubusercontent.com/15730971/229883580-da03461c-55c3-4404-adef-7f10e8dcc3b6.png">

2. There's a divergence when it comes to product reviews when comparing the Single Product block alongside the default Single Product Template. Is that expected?

<img width="596" alt="Screenshot 2023-04-04 at 15 28 31" src="https://user-images.githubusercontent.com/15730971/229885504-257ef3c6-baff-44a3-9aa0-858c7b372e09.png">

3. The SKU block is not being rendered on the FE.

nefeline commented on src/BlockTypes/SingleProduct.php:
```diff
@@ -0,0 +1,152 @@
+			$block_name = array_pop( $this->single_product_inner_blocks_names );
+
+			if ( $block_name === $block['blockName'] ) {
+				if ( 'core/post-excerpt' === $block_name ) {
```
What do you think about adding a comment here to document that this is a temporary overwrite until https://github.com/WordPress/gutenberg/pull/49495 is merged?

thealexandrelara replied:
Totally agree with you, I'll be adding a comment there

thealexandrelara replied:
Done! 🙌

thealexandrelara committed:
Fix error 'Block names must be string' when adding Single Product block

thealexandrelara commented:
Thank you @Aljullu  and @nefeline for reviewing it again and sharing the issues that you've found

> For some reason, the Product Title isn't showing up for me when it's before the Product Summary block. Can you reproduce as well?

Unfortunately I can't reproduce it, I tried different scenarios: on the Home template, on the Single Product template, on a post and on a Page, and all of them are rendering the product title when I move the Product Summary bellow it

> When adding the Single Product block to a template, on the editor side, I get the following error: Block names must be strings.

Ohh nice catch! I just pushed a fix to it, the Product Reviews block was not following the new `registerBlockSingleProductTemplate` method signature

> There's a divergence when it comes to product reviews when comparing the Single Product block alongside the default Single Product Template. Is that expected?

I think it should be displaying the reviews count too, I'll open an issue to start working on that

> The SKU block is not being rendered on the FE

This is being fixed by Luigi on [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/8805)

thealexandrelara committed:
Add todo comment to state the temporary solution used in the Single Product block

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-block-server-side

Aljullu reviewed (approved):
From my side, I think we can merge this PR as-is and work in the other issues in follow-ups. Good job, @thealexandrelara, this is an exciting achievement! :clap: Leaving it to @nefeline to do the final approval, though.

Only a heads-up that I'm still able to reproduce this one: `Not sure if it needs to be part of this PR, but mentioning in case it's not known: the Product Meta block doesn't seem to work on the Single Product block`. I think there is no issue for it, so it might be good to create one so we don't forget.

> Unfortunately I can't reproduce it, I tried different scenarios: on the Home template, on the Single Product template, on a post and on a Page, and all of them are rendering the product title when I move the Product Summary bellow it

This only happens to me when GB is disabled and in WP 6.2. But if that's only occurring to me, feel free to merge and I can investigate it later.

nefeline commented on src/BlockTypes/SingleProduct.php:
```diff
@@ -0,0 +1,153 @@
+					setup_postdata( $post );
+				}
+				$context['postId'] = $this->product_id;
+				if ( ! $this->single_product_inner_blocks_names ) {
```
Do you mind clarifying the logic here? This condition is currently inside another `if ( $this->single_product_inner_blocks_names )` so I don't think there will ever be a case where this condition will be true.

thealexandrelara replied:
Sure! The idea is that I'm using `array_pop` on line 125 to keep getting each Inner Block name until the array is empty. The `render_block_context` is called once for each block in a sequential way, so after iterating over all the inner blocks, the array should be empty, and then the `wp_reset_postdata` is called.

Now that you mention I think it could be better if we move this condition outside the `if ( $block_name === $block['blockName'] ) ` so we can leave it at the same level of the `array_pop`. I just did this change!

nefeline replied:
> Sure! The idea is that I'm using array_pop on line 125 to keep getting each Inner Block name until the array is empty. The render_block_context is called once for each block in a sequential way, so after iterating over all the inner blocks, the array should be empty, and then the wp_reset_postdata is called.

> Now that you mention I think it could be better if we move this condition outside the if ( $block_name === $block['blockName'] )  so we can leave it at the same level of the array_pop. I just did this change!

Thanks for clarifying & the update! Perhaps running `wp_reset_postdata()` by the end of every render of this block could be avoided, but since it's behind the feature flag and is also temporary, all good 👍

nefeline reviewed (commented):
Thanks for addressing all feedback! I did another round of tests; there's only one problem that surfaced that I would recommend fixing before merging, as it is a regression when using the post excerpt block outside of the single product block. The following example is when adding this block to the single product template. On the editor, the following error is triggered:

<img width="1233" alt="Screenshot 2023-04-06 at 10 24 50" src="https://user-images.githubusercontent.com/15730971/230397917-d834a3a7-cf60-43ac-a46e-f293f525169d.png">

thealexandrelara commented:
> From my side, I think we can merge this PR as-is and work in the other issues in follow-ups. Good job, @thealexandrelara, this is an exciting achievement! 👏 Leaving it to @nefeline to do the final approval, though.
> 
> Only a heads-up that I'm still able to reproduce this one: `Not sure if it needs to be part of this PR, but mentioning in case it's not known: the Product Meta block doesn't seem to work on the Single Product block`. I think there is no issue for it, so it might be good to create one so we don't forget.
> 
> > Unfortunately I can't reproduce it, I tried different scenarios: on the Home template, on the Single Product template, on a post and on a Page, and all of them are rendering the product title when I move the Product Summary bellow it
> 
> This only happens to me when GB is disabled and in WP 6.2. But if that's only occurring to me, feel free to merge and I can investigate it later.

The Product Meta block is being fixed by Luigi in [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/8966). I'll investigate more about this issue that you commented

Thank you so much for the reviews, insights and suggestions 🙌

thealexandrelara commented:
> Thanks for addressing all feedback! I did another round of tests; there's only one problem that surfaced that I would recommend fixing before merging, as it is a regression when using the post excerpt block outside of the single product block. The following example is when adding this block to the single product template. On the editor, the following error is triggered:
> 
> <img alt="Screenshot 2023-04-06 at 10 24 50" width="1233" src="https://user-images.githubusercontent.com/15730971/230397917-d834a3a7-cf60-43ac-a46e-f293f525169d.png">

Thank you for raising this @nefeline, this is an error that was occurring in Gutenberg, but they solved it last week: https://github.com/WordPress/gutenberg/pull/48730

thealexandrelara committed:
Move the wp_reset_postdata outside the block name check condition

thealexandrelara committed:
Add comments about why we need to unregister the block on Single Product template

nefeline reviewed (approved):
> Thank you for raising this @nefeline, this is an error that was occurring in Gutenberg, but they solved it last week: https://github.com/WordPress/gutenberg/pull/48730

Gotcha! Thanks for sharing this info. The curious thing is that the error only occurs when the excerpt is used outside of the single product block; perhaps something to explore later on.

Thanks again for all your work here! I'm a-ok with merging this PR as-is and working on other issues in follow-ups as well. It's fantastic to see this block in place: keep it up! 🚢

thealexandrelara commented:
> > Thank you for raising this @nefeline, this is an error that was occurring in Gutenberg, but they solved it last week: [WordPress/gutenberg#48730](https://github.com/WordPress/gutenberg/pull/48730)
> 
> Gotcha! Thanks for sharing this info. The curious thing is that the error only occurs when the excerpt is used outside of the single product block; perhaps something to explore later on.
> 
> Thanks again for all your work here! I'm a-ok with merging this PR as-is and working on other issues in follow-ups as well. It's fantastic to see this block in place: keep it up! 🚢

Thank you so much for providing detailed reviews and suggestions! Looking forward to see your PR getting merged on Gutenberg, so we can remove this temporary solution and replace it with the suggestion that you made!

opr commented:
@thealexandrelara I have added the correct labels to this PR.

opr commented:
I am skipping adding this to the testing notes as it is experimental. I will check the box in the PR body.

easilyamused commented:
I'm not seeing the block in the released package version 10.0.0 from the .org repo.

Aljullu commented:
Hey there @easilyamused, sorry for the confusion, this block is still marked as experimental and hasn't been released for public use yet. There are still some improvements and fixes we would like to introduce before making the Single Product block available to everybody. If you want to test it now, you can do so [downloading one of our nightly builds](https://github.com/woocommerce/woocommerce-blocks/releases). However, keep in mind it's experimental so it's not recommended to use it on production sites.

We have updated the changelog to make sure this block isn't mentioned there. Hopefully, in the following weeks we will be able to graduate it and make it available in the plugin. :slightly_smiling_face:
